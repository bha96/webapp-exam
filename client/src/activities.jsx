import { useEffect, useState } from "react";

export function Activities() {
  const [activities, setActivities] = useState([]);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/activities");

      if (res.ok) {
        setActivities(await res.json());
      } else {
        throw await res.body;
      }
    })();
  }, []);

  return (
    <div>
      {activities.map((a) => (
        <h1 key={a.name}>{a.name}</h1>
      ))}
    </div>
  );
}
