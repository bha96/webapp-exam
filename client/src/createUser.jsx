import { useState } from "react";

export function CreateUser() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fullName, setFullName] = useState("");
  const [alert, setAlert] = useState("");
  const [color, setColor] = useState("black");

  async function handleSubmit(e) {
    e.preventDefault();
    console.log(username, fullName, password);

    const res = await fetch("/api/users/new", {
      method: "POST",
      body: JSON.stringify({ username, password, fullName }),
      headers: {
        "content-type": "application/json",
      },
    });

    if (res.ok) {
      setColor("green");
      setAlert("User created");
    } else if (res.status === 409) {
      setColor("red");
      setAlert("User already exists");
    }
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h1>Create a new user</h1>
        <p>Everything is case sensitive</p>
        <div>
          <label>
            Username:
            <input
              type="text"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
              placeholder={"Squirrel Man"}
              required={true}
            />
          </label>
        </div>
        <div>
          <label>
            Full name:
            <input
              type="text"
              value={fullName}
              onChange={(e) => setFullName(e.target.value)}
              placeholder={"T.R. Squirrlington"}
              required={true}
            />
          </label>
        </div>
        <div>
          <label>
            Password:
            <input
              type="password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              placeholder={"******"}
              required={true}
            />
          </label>
        </div>

        <div>User will automatically be assigned to your group</div>
        <button>Create user</button>
      </form>
      <div style={{ color: color }}>{alert}</div>
    </div>
  );
}
