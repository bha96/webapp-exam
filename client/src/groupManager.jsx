import { useEffect, useState } from "react";

export function GroupManager(mongoDatabase) {
  const [users, setUsers] = useState([]);
  useEffect(() => {
    (async () => {
      const res = await fetch("/api/users");
      if (res.ok) {
        setUsers(await res.json());
      }
    })();
  }, []);

  return (
    <div>
      {users.map((u) => (
        <div>{u.fullName}</div>
      ))}
    </div>
  );
}
