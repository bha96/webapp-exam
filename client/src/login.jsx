import { useState } from "react";
import { useNavigate } from "react-router-dom";

export function Login({ setUser }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [alert, setAlert] = useState("");
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();

    const res = await fetch("api/login", {
      method: "POST",
      body: JSON.stringify({ username, password }),
      headers: {
        "content-type": "application/json",
      },
    });

    if (res.ok) {
      const user = await res.json();
      setUser(user);
      console.log(user.userType);
      if (user.userType === "manager") {
        navigate("/group-manager");
      } else {
        navigate("/");
      }
    } else {
      setAlert("Wrong username or password");
    }
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h2>Login</h2>
        <div>
          <label>
            Username:
            <input
              type="text"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
            />
          </label>
        </div>
        <div>
          <label>
            Password:
            <input
              type="password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
          </label>
        </div>
        <button>Log in</button>
      </form>
      <p style={{ color: "red" }}>{alert}</p>
    </div>
  );
}
