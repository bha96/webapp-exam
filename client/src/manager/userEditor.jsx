import * as React from "react";
import { useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate } from "react-router-dom";

export function UserEditCard({ user, setChanged }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fullName, setFullName] = useState("");
  const [alert, setAlert] = useState("");
  const navigate = useNavigate();

  async function handleUsername(e) {
    e.preventDefault();
    const res = await fetch(`/api/users/${user._id}`, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ username }),
    });
    if (res.ok) {
      console.log("All good");
    } else if (res.status === 409) {
      setAlert("Username already exists");
      setChanged((old) => !old);
    } else {
      console.log("Something went awry");
    }
  }

  async function handlePassword(e) {
    e.preventDefault();
    const res = await fetch(`/api/users/${user._id}`, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ password }),
    });
    if (res.ok) {
      console.log("All good");
      setChanged((old) => !old);
    } else {
      console.log("Something went 2");
    }
  }

  async function handleFullName(e) {
    e.preventDefault();
    const res = await fetch(`/api/users/${user._id}`, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ fullName }),
    });
    if (res.ok) {
      console.log("All good");
      setChanged((old) => !old);
      navigate(`../${fullName}`);
    } else {
      console.log("Something went awry");
    }
  }

  async function deleteUser() {
    const res = await fetch(`/api/users/${user.username}`, {
      method: "DELETE",
    });
    if (res.ok) {
      console.log("All good");
      navigate("..");
      setChanged((old) => !old);
    } else {
      console.log("Something went wrong");
    }
  }

  return (
    <div>
      <h1>{user.fullName}</h1>
      <form onSubmit={handleUsername}>
        <label>
          Username:
          <input
            type="text"
            value={username}
            onChange={(e) => setUsername(e.target.value)}
            placeholder={user.username}
          />
        </label>
        <button>Change</button>
      </form>
      <form onSubmit={handlePassword}>
        <label>
          Password:&nbsp;
          <input
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            placeholder={"*****"}
          />
        </label>
        <button>Change</button>
      </form>
      <form onSubmit={handleFullName}>
        <label>
          Full name:
          <input
            type="text"
            value={fullName}
            onChange={(e) => setFullName(e.target.value)}
            placeholder={user.fullName}
          />
        </label>
        <button>Change</button>
      </form>
      <button onClick={deleteUser}>Delete user</button>
      {alert}
    </div>
  );
}

export function UserEditor() {
  const [users, setUsers] = useState([]);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/users");
      if (res.ok) {
        setUsers(await res.json());
      } else {
        console.log("Something went wrong " + res.statusMessage);
      }
    })();
  }, [changed]);

  return (
    <div>
      <h1>Edit users</h1>
      {users.map((u) => (
        <div key={u._id}>
          <Link to={`${u.fullName}`}>{u.fullName}</Link>
        </div>
      ))}
      <Routes>
        {users.map((u) => (
          <Route
            key={u._id}
            path={`/${u.fullName}`}
            element={
              <UserEditCard
                user={u}
                changed={changed}
                setChanged={setChanged}
              />
            }
          />
        ))}
      </Routes>
    </div>
  );
}
