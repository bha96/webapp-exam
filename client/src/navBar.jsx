import * as React from "react";
import { Link, useNavigate } from "react-router-dom";

export function NavBar({ user, setUser }) {
  const navigate = useNavigate();
  function logOut() {
    console.log("Logging out");
    fetch("/api/login", {
      method: "DELETE",
    }).then((r) => console.log(r));
    setUser({});
    navigate("/login");
  }

  return (
    <nav className={"navBar"}>
      {user.userType === "manager" && (
        <Link to={"/group-manager"}>Manage your group</Link>
      )}

      {user.userType === "user" && <Link to={"/activities"}>Activities</Link>}
      {user.username && <div>Greetings {user.fullName}!</div>}
      {user.username && (
        <button onClick={logOut} type={"button"}>
          Log out
        </button>
      )}
    </nav>
  );
}
