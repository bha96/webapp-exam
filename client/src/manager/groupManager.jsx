import { Link, Route, Routes } from "react-router-dom";
import { TeamChooser } from "./teamChooser";
import { CreateUser } from "./createUser";
import { UserEditor } from "./userEditor";
import { ActivityEditor } from "./activityEditor";
import * as React from "react";

export function GroupManager({ user }) {
  return (
    <div>
      <h1>"Company logo"</h1>
      <Link to={"./team-chooser"}>Team Chooser</Link>
      &nbsp;&nbsp;
      <Link to={"./users/new"}>Create user</Link>
      &nbsp;&nbsp;
      <Link to={"./users/edit"}>Edit users</Link>
      &nbsp;&nbsp;
      <Link to={"./activities/edit"}>Edit activities</Link>
      <Routes>
        <Route path={"/team-chooser"} element={<TeamChooser user={user} />} />
        <Route path={"/users/new"} element={<CreateUser />} />
        <Route path={"/users/edit/*"} element={<UserEditor />} />
        <Route path={"/activities/edit/*"} element={<ActivityEditor />}></Route>
      </Routes>
    </div>
  );
}
