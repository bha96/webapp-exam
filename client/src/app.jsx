import { BrowserRouter, Route, Routes, useNavigate } from "react-router-dom";
import { NavBar } from "./navBar";

import { Login } from "./login";
import { NotFound } from "./notFound";
import * as React from "react";
import { useEffect, useState } from "react";
import { Activities } from "./activities";

function Home({ user }) {
  const navigate = useNavigate();
  useEffect(() => {
    return () => {
      if (!user.username) {
        navigate("/login");
      }
      if (user.username) {
        navigate("/activities");
      }
    };
  }, []);
}

export function App() {
  const [user, setUser] = useState({});
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/login");

      if (res.ok) {
        setUser(await res.json());
      }
      setLoading(false);
    })();
  }, []);

  if (loading) {
    return "Loading...";
  }

  return (
    <BrowserRouter>
      <NavBar user={user} setUser={setUser} />
      <Routes>
        <Route path={"/"} element={<Home user={user} />}></Route>
        {!user.username && (
          <Route path={"/login"} element={<Login setUser={setUser} />}></Route>
        )}
        {user.username && (
          <Route path={"/activities"} element={<Activities />}></Route>
        )}

        <Route path={"/*"} element={<NotFound />}></Route>
      </Routes>
    </BrowserRouter>
  );
}
