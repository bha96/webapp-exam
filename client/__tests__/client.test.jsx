import * as React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { Login } from "../src/login";
import { NotFound } from "../src/notFound";
import { NavBar } from "../src/navBar";
import { App } from "../src/app";
import { Activities, ActivityCard } from "../src/activities";
import { UserEditCard, UserEditor } from "../src/manager/userEditor";
import { TeamChooser, UserCard } from "../src/manager/teamChooser";
import { GroupManager } from "../src/manager/groupManager";
import { CreateUser } from "../src/manager/createUser";
import {
  ActivityEditCard,
  ActivityEditor,
} from "../src/manager/activityEditor";
describe("renders components", function () {
  global.fetch = jest.fn(() => {
    return {
      ok: true,
      json: () => {
        return [
          {
            key: 123,
            name: "Saving the world",
            group: [1, 2],
            hours: [{ hours: 0, username: "clark" }],
          },
        ];
      },
    };
  });

  it("shows login page", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("show not found page", function () {
    const component = renderer.create(<NotFound />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows navigation bar", async function () {
    let component;
    await act(
      async () =>
        (component = renderer
          .create(
            <MemoryRouter>
              <NavBar user={{ userType: "user", username: "test" }} />
            </MemoryRouter>
          )
          .toJSON())
    );
    expect(component).toMatchSnapshot();
  });

  it("shows front page", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <App />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows activities", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <Activities />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows activity cards", async function () {
    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [
            {
              key: 123,
              name: "Saving the world",
              group: [1, 2],
              hours: [{ hours: 0, username: "clark" }],
            },
          ];
        },
      };
    });

    let component;
    await act(
      async () =>
        (component = renderer
          .create(
            <MemoryRouter>
              <ActivityCard
                activity={{
                  name: "Saving the world",
                  group: [1, 2],
                  hours: [{ hours: 0, username: "clark" }],
                }}
              />
            </MemoryRouter>
          )
          .toJSON())
    );

    expect(component).toMatchSnapshot();
  });

  it("shows user editor", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <UserEditor />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it("shows user editor card", async function () {
    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [
            {
              key: 123,
              name: "Saving the world",
              group: [1, 2],
              hours: [{ hours: 0, username: "clark" }],
            },
          ];
        },
      };
    });
    let component;
    await act(
      async () =>
        (component = renderer
          .create(
            <MemoryRouter>
              <UserEditCard
                user={{
                  _id: "63f346e92ff514f3a1d7eaa1",
                  username: "carl",
                  password: "123",
                  fullName: "Carl Sagan",
                  userType: "user",
                  group: 1,
                  totalHours: 0,
                }}
              />
            </MemoryRouter>
          )
          .toJSON())
    );
    expect(component).toMatchSnapshot();
  });
  it("shows user team chooser", function () {
    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [
            {
              _id: "63f346e92ff514f3a1d7eaa1",
              username: "carl",
              password: "123",
              fullName: "Carl Sagan",
              userType: "user",
              group: 1,
              totalHours: 0,
            },
            {
              _id: "63f348402ff514f3a1d7eaa3",
              username: "george",
              password: "123",
              fullName: "George Clooney",
              userType: "user",
              group: 2,
              totalHours: 0,
            },
          ];
        },
      };
    });
    const component = renderer
      .create(
        <MemoryRouter>
          <TeamChooser
            user={{
              _id: "63f348402ff514f3a1d7eaa3",
              username: "george",
              password: "123",
              fullName: "George Clooney",
              userType: "user",
              group: 2,
              totalHours: 0,
            }}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows user card", async function () {
    let component;
    await act(
      async () =>
        (component = renderer
          .create(
            <MemoryRouter>
              <UserCard
                user={{
                  _id: "63f346e92ff514f3a1d7eaa1",
                  username: "carl",
                  password: "123",
                  fullName: "Carl Sagan",
                  userType: "user",
                  group: 1,
                  totalHours: 0,
                }}
              />
            </MemoryRouter>
          )
          .toJSON())
    );
    expect(component).toMatchSnapshot();
  });

  it("shows user GroupManager", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <GroupManager />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows user create user page", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <CreateUser />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows user activity editor", function () {
    const component = renderer
      .create(
        <MemoryRouter>
          <ActivityEditor />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("shows user activitiy edit card", async function () {
    let component;
    await act(
      async () =>
        (component = renderer
          .create(
            <MemoryRouter>
              <ActivityEditCard
                activity={{
                  name: "Saving the world",
                  group: [1, 2],
                  hours: [{ hours: 0, username: "clark" }],
                }}
              />
            </MemoryRouter>
          )
          .toJSON())
    );
    expect(component).toMatchSnapshot();
  });
});
