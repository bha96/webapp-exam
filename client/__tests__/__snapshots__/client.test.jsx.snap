// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders components show not found page 1`] = `
<div>
  <h1>
    404:
  </h1>
  <h3>
    This is not the page you are looking for
  </h3>
</div>
`;

exports[`renders components shows activities 1`] = `
<div>
  <hr />
  <div />
</div>
`;

exports[`renders components shows front page 1`] = `"Loading..."`;

exports[`renders components shows login page 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Login
    </h2>
    <div>
      <label>
        Username:
        <input
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Password:
        <input
          onChange={[Function]}
          type="password"
          value=""
        />
      </label>
    </div>
    <button>
      Log in
    </button>
  </form>
  <p
    style={
      {
        "color": "red",
      }
    }
  />
</div>
`;

exports[`renders components shows navigation bar 1`] = `null`;

exports[`renders components shows user GroupManager 1`] = `
<div>
  <h1>
    "Company logo"
  </h1>
  <a
    href="/team-chooser"
    onClick={[Function]}
  >
    Team Chooser
  </a>
    
  <a
    href="/users/new"
    onClick={[Function]}
  >
    Create user
  </a>
    
  <a
    href="/users/edit"
    onClick={[Function]}
  >
    Edit users
  </a>
    
  <a
    href="/activities/edit"
    onClick={[Function]}
  >
    Edit activities
  </a>
</div>
`;

exports[`renders components shows user activitiy edit card 1`] = `null`;

exports[`renders components shows user activity editor 1`] = `
<div>
  <h1>
    Activity editor
  </h1>
</div>
`;

exports[`renders components shows user card 1`] = `null`;

exports[`renders components shows user create user page 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <h1>
      Create a new user
    </h1>
    <p>
      Everything is case sensitive
    </p>
    <div>
      <label>
        Username:
        <input
          onChange={[Function]}
          placeholder="Squirrel Man"
          required={true}
          type="text"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Full name:
        <input
          onChange={[Function]}
          placeholder="T.R. Squirrlington"
          required={true}
          type="text"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Password:
        <input
          onChange={[Function]}
          placeholder="******"
          required={true}
          type="password"
          value=""
        />
      </label>
    </div>
    <div>
      User will automatically be assigned to your group
    </div>
    <button>
      Create user
    </button>
  </form>
  <div
    style={
      {
        "color": "black",
      }
    }
  />
</div>
`;

exports[`renders components shows user editor 1`] = `
<div>
  <h1>
    Edit users
  </h1>
</div>
`;

exports[`renders components shows user editor card 1`] = `null`;

exports[`renders components shows user team chooser 1`] = `
<div>
  <h2>
    People on your team
  </h2>
  <hr />
  <h2>
    People not on your team
  </h2>
  <hr />
  <button
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
